### PR TITLE
📦 NEW: Port mg-policy to hardtanh ctx arch

### PR DESCRIPTION
**Pull request by ianagbip1oti**
```
-------------------------------------------------
Results of princhess vs princhess-main (40+0.4, 1t, 128MB, UHO_Lichess_4852_v1.epd):
Elo: 4.32 +/- 3.35, nElo: 8.09 +/- 6.26
LOS: 99.43 %, DrawRatio: 50.19 %, PairsRatio: 1.09
Games: 11822, Wins: 2580, Losses: 2433, Draws: 6809, Points: 5984.5 (50.62 %)
Ptnml(0-2): [64, 1343, 2967, 1456, 81], WL/DD Ratio: 0.48
LLR: 2.90 (100.4%) (-2.25, 2.89) [0.00, 3.00]
--------------------------------------------------

-------------------------------------------------
Results of princhess vs princhess-main (40+0.4, 1t, 128MB, DFRC_4852_v1.epd):
Elo: 3.41 +/- 4.41, nElo: 5.98 +/- 7.73
LOS: 93.50 %, DrawRatio: 47.91 %, PairsRatio: 1.06
Games: 7752, Wins: 1770, Losses: 1694, Draws: 4288, Points: 3914.0 (50.49 %)
Ptnml(0-2): [74, 907, 1857, 945, 93], WL/DD Ratio: 0.52
LLR: 2.91 (100.6%) (-2.25, 2.89) [-3.00, 0.00]
--------------------------------------------------

-------------------------------------------------
Results of princhess vs princhess-main (8+0.08, 1t, 128MB, UHO_Lichess_4852_v1.epd):
Elo: 3.99 +/- 4.66, nElo: 6.91 +/- 8.06
LOS: 95.35 %, DrawRatio: 47.83 %, PairsRatio: 1.07
Games: 7138, Wins: 1672, Losses: 1590, Draws: 3876, Points: 3610.0 (50.57 %)
Ptnml(0-2): [79, 819, 1707, 869, 95], WL/DD Ratio: 0.56
LLR: 2.92 (100.9%) (-2.25, 2.89) [-3.00, 0.00]
--------------------------------------------------
```

## Summary by CodeRabbit

- **New Features**
  - Introduced an updated move-scoring policy with context-aware evaluation and separate handling for favorable captures.
  - Added richer training reports, including per-piece accuracy, loss, information gain, gradient, and learning-rate metrics.
  - Training metadata and final analysis are now saved in TOML and binary formats.

- **Improvements**
  - Training uses batched gradient processing and improved smoothing and warmup analysis.
  - Quantized policy evaluation now supports expanded attention capacity and context-based scoring.

### CODE REVIEW

File: crates/princhess-train/src/bin/train-mg-policy.rs [hunk 1]
```diff
@@ -427,6 +472,311 @@ fn run_training_loop<S: LRScheduler + Sync>(
...
+        let _ = file.write_all(bytemuck::cast_slice(&lr_full));
+        let _ = file.write_all(bytemuck::cast_slice(&grad_l1));
+    }
+}
+
+fn write_training_toml(dir: &Path, sb: usize, stats: &TrainingStats, config: &TrainingConfig) {
+    let (loss, accuracy) = stats.get_prev_avg_metrics();
+    let info_gain = stats.get_prev_info_gain();
+    let (piece_accuracy, piece_info_gain, wrong_piece, wrong_square) = stats.get_prev_piece_data();
+    let piece_names = ["p", "n", "b", "r", "q", "k"];
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/train-mg-policy.rs:497]: piece_names is deliberately positional — the PNBRQK ordering is the canonical piece discriminant order throughout the engine, and any reordering would break far more than TOML labels. The array length is implicitly enforced: if Piece::COUNT changed, the enumerate() loop would produce wrong indices and the mismatch would surface immediately in tests. We're accepting this as-is.

File: src/quantized_mg_policy.rs [hunk 2]
```diff
@@ -1,38 +1,71 @@
...
 type FeatureVector = ArrayVec<usize, 32>;
 
+#[repr(C)]
+#[derive(Copy, Clone, Zeroable)]
+pub struct QuantizedMgCtxNetwork {
+    weights: [Align16<Accumulator<i16, CTX_SIZE>>; INPUT_SIZE],
+    bias: Align16<Accumulator<i16, CTX_SIZE>>,
+}
+
+unsafe impl Pod for QuantizedMgCtxNetwork {}
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply src/quantized_mg_policy.rs:42]: Addressed in https://github.com/princesslana/princhess/pull/482/commits/7f9711b72da8f862f7c96b8086d0d7f86facefa5

File: crates/princhess-train/src/bin/train-mg-policy.rs [hunk 3]
```diff
@@ -1078,22 +1431,30 @@ fn update_gradient(
...
     metrics.loss += position_loss;
     metrics.baseline_loss += baseline;
 
+    for idx in 0..moves.len() {
+        let piece = move_idxes[idx].piece();
+        let t_i = expected_primary[idx];
+        let log_p_i = actual_policy[idx].max(EPSILON).ln();
+        metrics.piece_loss[piece] += t_i * (-log_p_i);
+        metrics.piece_baseline_loss[piece] += t_i * (moves.len() as f32).ln();
+    }
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/train-mg-policy.rs:1202]: Addressed in https://github.com/princesslana/princhess/pull/482/commits/7f9711b72da8f862f7c96b8086d0d7f86facefa5
